### PR TITLE
Enable Backup Create/Delete for CSI Filestore driver

### PR DESCRIPTION
**Pull request by saikat-royc**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Commit1: This patch implements CreateSnapshot and DeleteSnapshot CSI API, for Filestore backups. The patch also generates a go client for v1beta1 filestore [apis](https://cloud.google.com/filestore/docs/reference/rest/v1beta1/projects.locations.backups) which implement backups.
Commit2: Support for creation of a new instance from a given snapshot source

**Does this PR introduce a user-facing change?**:

```release-note
Enable Backup Create/Delete  and Create Volume for a Backup source for CSI Filestore driver
```

### CODE REVIEW

**Comment by saikat-royc** [comment]
I will update the README in the same commit after addressing review comments

File: pkg/csi_driver/controller.go [hunk 1]
```diff
@@ -106,13 +106,36 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
...
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-
 	glog.V(5).Infof("Using capacity bytes %q for volume %q", capBytes, name)
 	newFiler, err := s.generateNewFileInstance(name, capBytes, req.GetParameters(), req.GetAccessibilityRequirements())
 	if err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
+	// TODO: The workflow to provision a volume from a volume snapshot source needs to be implemented.
```
Comment (mattcary) [inline pkg/csi_driver/controller.go:115]: Even as alpha I don't think we want it to be in a state of partial support. I'm happy to break it up into multiple commits/PR but I don't want to defer it with a TODO.

Comment (saikat-royc) [reply pkg/csi_driver/controller.go:115]: Right I am working on it and plan to submit separate commit in this PR

Comment (mattcary) [reply pkg/csi_driver/controller.go:115]: 👍🏼

File: pkg/csi_driver/controller.go [hunk 2]
```diff
@@ -483,3 +531,109 @@ func mergeLabels(scLabels map[string]string, metedataLabels map[string]string) (
...
+	}
+	filer.Project = s.config.metaService.GetProject()
+	// If parameters are empty we assume 'backup' type by default.
+	if req.GetParameters() != nil {
+		if _, err := util.IsSnapshotTypeSupported(req.GetParameters()); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	// Check for exisitng snapshot
```
Comment (mattcary) [inline pkg/csi_driver/controller.go:557]: exisitng typo

File: examples/kubernetes/backups-restore/sc.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+apiVersion: storage.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (saikat-royc) [reply examples/kubernetes/backups-restore/sc.yaml:1]: Let's take it up as a separate PR.

File: examples/kubernetes/backups-restore/sc.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+kind: StorageClass
+metadata:
+  name: csi-filestore
+provisioner: filestore.csi.storage.gke.io
+parameters:
+  labels: key1=value1,key2=value2
```
[the remark this answers is not in the document]
Comment (saikat-royc) [reply examples/kubernetes/backups-restore/sc.yaml:7]: Done

File: pkg/cloud_provider/file/file.go [hunk 5]
```diff
@@ -53,24 +55,37 @@ type Network struct {
...
 type Service interface {
 	CreateInstance(ctx context.Context, obj *ServiceInstance) (*ServiceInstance, error)
 	DeleteInstance(ctx context.Context, obj *ServiceInstance) error
 	GetInstance(ctx context.Context, obj *ServiceInstance) (*ServiceInstance, error)
 	ListInstances(ctx context.Context, obj *ServiceInstance) ([]*ServiceInstance, error)
 	ResizeInstance(ctx context.Context, obj *ServiceInstance) (*ServiceInstance, error)
+	GetBackup(ctx context.Context, backupUri string) (*BackupInfo, error)
+	CreateBackup(ctx context.Context, obj *ServiceInstance, backupId, backupLocation string) (*filev1beta1.Backup, error)
+	DeleteBackup(ctx context.Context, backupId string) error
+	CreateInstanceFromSnapshotSource(ctx context.Context, obj *ServiceInstance, volumeSourceSnapshotId string) (*ServiceInstance, error)
```
[the remark this answers is not in the document]
Comment (saikat-royc) [reply pkg/cloud_provider/file/file.go:72]: Done

File: pkg/csi_driver/controller.go (hunk 2 above)
Comment (saikat-royc) [reply pkg/csi_driver/controller.go:557]: Done

File: examples/kubernetes/backups-restore/sc.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattcary) [reply examples/kubernetes/backups-restore/sc.yaml:1]: ok

**Pull request merged**